kmip/pie/client.py

```diff
@@ -0,0 +1,380 @@
+from kmip.pie import factory
+from kmip.pie import objects as pobjects
+
+from kmip.services.kmip_client import KMIPProxy
+
+
+class KmipClient:
+    """
+    A simplified PyKMIP client for conducting KMIP operations.
+
+    The KmipClient provides a simpler external interface for various KMIP
+    operations and composes the bulk of the PyKMIP Pie API. It wraps the
+    original KMIPProxy, reducing the boilerplate needed to deploy PyKMIP
+    in client applications.
+    """
+    def __init__(self,
+                 hostname='127.0.0.1',
+                 port=5696,
+                 cert=None,
+                 key=None,
+                 ca=None,
+                 ssl_version=ssl.PROTOCOL_SSLv23,
+                 username=None,
+                 password=None,
+                 config='client'):
+        """
+        Construct a KMIPClient.
+
+        Args:
+            hostname (string): The host or IP address of a KMIP appliance.
+                Optional, defaults to '127.0.0.1'.
+            port (int): The port number used to establish a connection to a
+                KMIP appliance. Usually 5696 for KMIP applications. Optional,
+                defaults to 5696.
+            cert (string): The path to the client's certificate. Optional,
+                defaults to None.
+            key (string): The path to the key for the client's certificate.
+                Optional, defaults to None.
+            ca (string): The path to the CA certificate used to verify the
+                server's certificate. Optional, defaults to None.
+            ssl_version (ssl constant): The ssl version to use for the
+                connection. Optional, defaults to ssl.PROTOCOL_SSLv23.
+            username (string): The username of the KMIP appliance account to
+                use for operations. Optional, defaults to None.
+            password (string): The password of the KMIP appliance account to
+                use for operations. Optional, defaults to None.
+            config (string): The name of a section in the PyKMIP configuration
+                file. Use to load a specific set of configuration settings from
+                the configuration file, instead of specifying them manually.
+                Optional, defaults to the default client section, 'client'.
+        """
+        self.logger = logging.getLogger()
+
+        self.attribute_factory = attributes.AttributeFactory()
+        self.object_factory = factory.ObjectFactory()
+
+        # TODO (peter-hamilton) Consider adding validation checks for inputs.
+        self.proxy = KMIPProxy(
+            host=hostname,
+            port=port,
+            certfile=cert,
+            keyfile=key,
+            ca_certs=ca,
+            ssl_version=ssl_version,
+            username=username,
+            password=password,
+            config=config)
+
+        self._is_open = False
+        self._lock = multiprocessing.RLock()
+
+    def _synchronize(f):
```

Comment:
I think that multi-threading support is critical, but I agree that it might not be necessary for the initial implementation. (Adding synchronization barriers would **not** change the existing interface.)

If we're going to include synchronization, let's use an existing library (e.g., [oslo.concurrency.lockutils](http://docs.openstack.org/developer/oslo.concurrency/api/lockutils.html)) rather than implementing one ourselves. In the very least, the synchronization decorator belongs in a utils modules rather than here.